src/System.IO.Pipelines/src/System/IO/Pipelines/PipeAwaitable.cs

```diff
@@ -73,39 +88,54 @@ public void Reset()
             }
         }
 
-        public bool IsCompleted => ReferenceEquals(_state, _awaitableIsCompleted);
-        internal bool HasContinuation => !ReferenceEquals(_state, _awaitableIsNotCompleted);
+        public bool IsCompleted => ReferenceEquals(_completion, _awaitableIsCompleted);
+        internal bool HasContinuation => !ReferenceEquals(_completion, _awaitableIsNotCompleted);
 
-        public Action OnCompleted(Action continuation, out bool doubleCompletion)
+        public void OnCompleted(Action<object> continuation, object state, ValueTaskSourceOnCompletedFlags flags, out Action<object> completion, out object completionState, out bool doubleCompletion)
         {
+            completionState = null;
+            completion = null;
+
             doubleCompletion = false;
-            Action awaitableState = _state;
+            Action<object> awaitableState = _completion;
             if (ReferenceEquals(awaitableState, _awaitableIsNotCompleted))
             {
-                _state = continuation;
+                _completion = continuation;
+                _completionState = state;
+
+                if ((flags & ValueTaskSourceOnCompletedFlags.UseSchedulingContext) != 0)
```

Comment:
There's also FlowExecutionContext.  That doesn't matter when await'ing in a Task/ValueTask-returning async method, as those builders automatically flow ExecutionContext across awaits (and then use the awaiter's UnsafeOnCompleted to avoid duplicate flow).  But it would matter in two cases:
1. Someone uses the ValueTask's GetAwaiter().OnCompleted(...) directly rather than using await.  In that case by ignoring FlowExecutionContext here you won't end up propagating things like AsyncLocals.
2. Someone uses a custom async method builder (e.g. an async method returning something other than Task/ValueTask) and doesn't do the due diligence to propagate ExecutionContext across awaits.  In that case again, you won't end up propagating things like AsyncLocals.

Reply:
The problem is that I don’t want to complicate the scheduler abstraction with these flags. The 2 default scheduler implementations do capture the execution context and that’s why I ignored it.

To make this full fidelity, we need to pass it to the scheduler.

Reply:
> The 2 default scheduler implementations do capture the execution context and that’s why I ignored it.

If you're always flowing already, that's fine.  You'll just want to verify the capture actually happens at the right place, e.g. that you're calling ExecutionContext.Capture during the OnCompleted call rather than when the operation completes.

Reply:
Actually, thinking it about it I think we need to do something here and it'll force more allocations as well 😢 . Assuming these flags are set correctly when using the Task and ValueTask  async state machine so it's only for those narrow cases you describe.

How are these flags supposed to compose? I guess I can look at your PR 😄